core/templates/dev/head/pages/exploration-player-page/services/learner-answer-info.service.spec.ts

```diff
@@ -0,0 +1,337 @@
+              content_id: 'feedback_1',
+              html: ''
+            },
+            labelled_as_correct: false,
+            param_changes: [],
+            refresher_exploration_id: null,
+            missing_prerequisite_skill_id: null
+          },
+          rule_specs: [{
+            inputs: {
+              x: 10
+            },
+            rule_type: 'Equals'
+          }]
+        }, {
+          outcome: {
+            dest: 'outcome 2',
+            feedback: {
+              content_id: 'feedback_2',
+              html: ''
+            },
+            labelled_as_correct: false,
+            param_changes: [],
+            refresher_exploration_id: null,
+            missing_prerequisite_skill_id: null
+          },
+          rule_specs: [{
+            inputs: {
+              x: 5
+            },
+            rule_type: 'Equals'
+          }, {
+            inputs: {
+              x: 7
+            },
+            rule_type: 'NotEquals'
+          }, {
+            inputs: {
+              x: 6
+            },
+            rule_type: 'Equals'
+          }]
+        }],
+        default_outcome: {
+          dest: 'default',
+          feedback: {
+            content_id: 'default_outcome',
+            html: ''
+          },
+          labelled_as_correct: false,
+          param_changes: [],
+          refresher_exploration_id: null,
+          missing_prerequisite_skill_id: null
+        },
+        hints: []
+      },
+      param_changes: [],
+      solicit_answer_details: true,
+      written_translations: {
+        translations_mapping: {
+          content: {},
+          default_outcome: {},
+          feedback_1: {},
+          feedback_2: {}
+        }
+      }
+    };
+
+    sof = $injector.get('StateObjectFactory');
+    oof = $injector.get('OutcomeObjectFactory');
+    acrof = $injector.get('AnswerClassificationResultObjectFactory');
+    LearnerAnswerInfoService = $injector.get('LearnerAnswerInfoService');
+    ladbas = $injector.get(
+      'LearnerAnswerDetailsBackendApiService');
+    DEFAULT_OUTCOME_CLASSIFICATION = $injector.get(
+      'DEFAULT_OUTCOME_CLASSIFICATION');
+    state = sof.createFromBackendDict('new state', stateDict);
+    spyOn(
+      mockAnswerClassificationService,
+      'getMatchingClassificationResult').and.returnValue(acrof.createNew(
+      oof.createNew('default', 'default_outcome', '', []), 2, 0,
+      DEFAULT_OUTCOME_CLASSIFICATION));
+    // Spying the random function to return 0, so that
+    // getRandomProbabilityIndex() returns 0, which is a private function in
+    // LearnerAnswerInfoService. This will help to mark the
+    // canAskLearnerAnswerInfo which is a boolean variable as true as every
+    // probability index is greater than 0.
+    spyOn(Math, 'random').and.returnValue(0);
+    LearnerAnswerInfoService.initLearnerAnswerInfoService(
+      '10', state, 'a', 'b');
+  }));
+
+  it('should return can ask learner for answer info true', function() {
```

Comment:
Tests are still weak. What about the cases where the learner has already been shown "answer info"? What about subsequent answers being submitted after initialization? Etc.

Please think through all the different _behaviours_ and contexts of your service and add tests accordingly. Don't just do a superficial poke at the various functions. @aks681 might be able to give you more advice on this.

Reply:
Made tests more robust